### DISCUSSION TITLE
Use Piranha's API without loading the manager

### DISCUSSION THREAD

**Discussion by gmullernh** [PiranhaCMS/piranha.core#1661] [Q&A] [answered]
Hello!

**TL;DR;**
Is possible to use the Piranha's API without loading the manager endpoints in an application?
I need to load some content (pages) from Piranha (read-only) and display them as JSON at specific endpoints. 
Would like to do this without starting the Piranha's manager endpoints. 

I'm using `services.AddPiranha` with `options.DisableRouting();` at `ConfigureServices` and without `app.UseManager()`, but the manager endpoints are being loaded.

I'm looking towards building a CMS environment, which consists of: 
- a instance of Piranha CMS that will manage user generated content and pages and write it to a database. (example.org/manager). [This one will use Piranha's Manager and it's already solved].
- an open API which will be consumed by external platforms (api.example.org). This API already exists and loads data from different sources. My idea here would be to "plug in" the Piranha's API, so we can load the different types of Pages (more specifically) from the database and deliver them as JSON.

We're splitting the concerns of the applications to improve security and performance.

I'm planning to share the data models and controller behavior between these applications with Piranha's Modules packages. Already tested and it does work pretty nicely. So, for each type of page/content, I'm creating a module (which give us good access control and isolation) and referencing it on both applications.

![Illustrative model](https://user-images.githubusercontent.com/30841186/126038219-20e897ae-2eaf-4534-84d1-291bcf0bd716.jpg)

How Piranha's API is being implemented at my consumer API (api.example.org) `Startup.cs`. 

```
	public void ConfigureServices(IServiceCollection services)
	{
            services.AddPiranha(options =>
            {
                options.DisableRouting();
                options.UseModuleA();
                options.UseEF<SQLServerDb>(...);
            });
	}


        public void Configure(IApplicationBuilder app, IWebHostEnvironment env, IApi api)
        {
            App.Init(api);

           // Noticed that Build() tries to save the changes to the database, is it possible to only read the content?
            new ContentTypeBuilder(api)
                .AddAssembly(typeof(ModuleA.Module).Assembly)
                .Build()
                .DeleteOrphans();

            app.UsePiranha(options =>
            {
                options.UseModuleA();
            });
        }
```

Even without the UseManager, after starting the application, Swagger shows all the /manager endpoints from Piranha.
![swagger](https://user-images.githubusercontent.com/30841186/126039425-de8c98f7-b649-4386-9e7f-b6ebda86ded0.png)

Is it possible through invocating or not invocating some methods or this would require overriding/rebuilding some stuff?
I'm trying to stay most "vanilla" as possible to allow for future updates and a smooth developer experience.
Best regards!

**Comment by tidyui** [comment, marked as answer]
Hi there! Asp.net registers all available controllers automatically. If you don’t want the manager endpoints or functionality, simply remove the NuGet package Piranha.Manager

best regards

**Reply by gmullernh** [reply]
Thank you for your quick reply!

It took me a while to realize.

I wasn't referencing the `Piranha.Manager` on the API, but it was being used by the **ModuleA** since it defines the Manager behavior (it was created using the default template). Since the module was being used by the API, the manager endpoints were being loaded.

My initial idea was that, if someone need a new data model, we could just create a new Module, define it's properties and behavior, how it would work in the manager (roles, UI and so on) and how it would be displayed by the API, through overriding the controllers. A magically neat plug'n'play packaging for each content. 

Solved by moving the **ModuleA** data model to a new class library project, using only `Piranha` and dedicated to the models. This project will be referenced by the modules and the API. The modules will handle Piranha CMS Manager behavior for each model, while the API will fetch the content (using Piranha's API) through it's own endpoints.

![new approach with data model layer](https://user-images.githubusercontent.com/30841186/126048955-a12f1895-5051-4dd7-8d9c-752891ceff9e.jpg)

Best regards!
